src/librustc_mir/transform/generator.rs

```diff
@@ -0,0 +1,728 @@
+            source_info,
+            kind: term,
+        }),
+        is_cleanup: false,
+    });
+}
+
+fn creator_generator_resume_function<'a, 'tcx>(
+        tcx: TyCtxt<'a, 'tcx, 'tcx>,
+        mut transform: TransformVisitor<'a, 'tcx>,
+        def_id: DefId,
+        source: MirSource,
+        mir: &mut Mir<'tcx>) {
+    // Poison the generator when it unwinds
+    for block in mir.basic_blocks_mut() {
+        let source_info = block.terminator().source_info;
+        if let &TerminatorKind::Resume = &block.terminator().kind {
+            block.statements.push(transform.set_state(1, source_info));
+        }
+    }
+
+    let source_info = SourceInfo {
+        span: mir.span,
+        scope: ARGUMENT_VISIBILITY_SCOPE,
+    };
+
+    let poisoned_block = BasicBlock::new(mir.basic_blocks().len());
+
+    let term = TerminatorKind::Assert {
+        cond: Operand::Constant(box Constant {
+            span: mir.span,
+            ty: tcx.types.bool,
+            literal: Literal::Value {
+                value: ConstVal::Bool(false),
+            },
+        }),
+        expected: true,
+        msg: AssertMessage::GeneratorResumedAfterPanic,
+        target: transform.return_block,
+        cleanup: None,
+    };
+
+    mir.basic_blocks_mut().push(BasicBlockData {
+        statements: Vec::new(),
+        terminator: Some(Terminator {
+            source_info,
+            kind: term,
+        }),
+        is_cleanup: false,
+    });
+
+    transform.bb_targets.insert((poisoned_block, None), 1);
+
+    let switch = TerminatorKind::SwitchInt {
+        discr: Operand::Consume(transform.make_field(transform.state_field, tcx.types.u32)),
+        switch_ty: tcx.types.u32,
+        values: Cow::from(transform.bb_targets.values().map(|&i| {
+                ConstInt::U32(i)
+            }).collect::<Vec<_>>()),
+        targets: transform.bb_targets.keys()
+            .map(|&(k, _)| k)
+            .chain(once(transform.return_block))
+            .collect(),
+    };
+
+    insert_entry_point(mir, BasicBlockData {
+        statements: Vec::new(),
+        terminator: Some(Terminator {
+            source_info,
+            kind: switch,
+        }),
+        is_cleanup: false,
+    });
+
+    make_generator_state_argument_indirect(tcx, def_id, mir);
+
+    no_landing_pads(tcx, mir);
+
+    // Make sure we remove dead blocks to remove
+    // unrelated code from the drop part of the function
+    simplify::remove_dead_blocks(mir);
+
+    dump_mir(tcx, None, "generator_resume", &0, source, mir);
+}
+
+fn insert_clean_drop<'a, 'tcx>(mir: &mut Mir<'tcx>) -> BasicBlock {
+    let source_info = SourceInfo {
+        span: mir.span,
+        scope: ARGUMENT_VISIBILITY_SCOPE,
+    };
+
+    let return_block = BasicBlock::new(mir.basic_blocks().len());
+    mir.basic_blocks_mut().push(BasicBlockData {
+        statements: Vec::new(),
+        terminator: Some(Terminator {
+            source_info,
+            kind: TerminatorKind::Return,
```

Comment:
Shouldn't this be a `TerminatorKind::GeneratorDrop`? I think we map `GeneratorDrop` for `Return`, but this is only a part of the "empty drop path".

Reply:
`Return` and `GeneratorDrop` has the same meaning for `create_generator_drop_shim`, it's just used as a distinction for earlier MIR passes.